p4-16/spec/P4-16-spec.mdk

```diff
@@ -4266,11 +4266,27 @@ h = { 10, 12 };  // This also makes the header h valid
 h = { y = 12, x = 10 };  // Same effect as above
 ~ End P4Example
 
-Two headers can be compared for equality (==) or inequality (!=) only
+Two headers can be compared for equality (`==`) or inequality (`!=`) only
 if they have the same type.  Two headers are equal if and only if they
 are both invalid, or they are both valid and all their corresponding
 fields are equal.
 
+The expression `(H) {#}` represents an invalid header of type `H`,
+where `H` is a header type.  For example:
+
+~ Begin P4Example
+header H { bit<32> x; bit<32> y; }
+H h;
+h = (H) {#};  // This makes the header h invalid
+if (h == (H) {#}) {  // This is equivalent to the condition !h.isValid()
+    // ...
+}
+~ End P4Example
+
+Note that the `#` character in this expression must not be the first
```

Comment:
the grammar should be inserted here.

Reply:
An alternative grammar is H.invalid

Reply:
Another alternative grammar is header.invalid

Reply:
`S s = { ... } => S s = { (H){#} };`

Reply:
H.invalid does not work so well for generic headers.

Reply:
Can we have implicit casting of invalid headers added, similar how we do implicit casting of `int` ? I think that in both examples, `(H)` can be be easily figured out from the context. Not a biggie, but would be nice and I think that will eliminate the need for explicit casts in most of the cases.

Reply:
Yes, we plan to do this in two steps. first require casting, then implement type inference and make casting optional.
We first want to implement fully the `...` syntax, though.
As always, I am leery putting something in the spec which isn't implemented yet.
The main question to discuss still is whether {#} is a single lexeme or 3.

Reply:
grammar inserted